File: ChatGPT(예진)/Java (수정 완료)/Stack_Manipulator.java
import java.util.Scanner;
import java.util.Stack;

class StackWithTimestamp {
    private Stack<ItemWithTimestamp> items;

    public StackWithTimestamp() {
        this.items = new Stack<>();
    }

    public void push(String item) {
        long timestamp = System.currentTimeMillis();
        items.push(new ItemWithTimestamp(item, timestamp));
    }

    public ItemWithTimestamp pop() {
        if (!isEmpty()) {
            return items.pop();
        }
        return null;
    }

    public boolean isEmpty() {
        return items.isEmpty();
    }

    public ItemWithTimestamp peek() {
        if (!isEmpty()) {
            return items.peek();
        }
        return null;
    }

    public static class ItemWithTimestamp {
        String item;
        long timestamp;

        public ItemWithTimestamp(String item, long timestamp) {
            this.item = item;
            this.timestamp = timestamp;
        }

        public String getItem() {
            return item;
        }

        public long getTimestamp() {
            return timestamp;
        }
    }
}

public class Stack_Manipulator {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        StackWithTimestamp stack = new StackWithTimestamp();
        StackWithTimestamp clipboard = new StackWithTimestamp();  // 클립보드 스택 생성

        while (true) {
            System.out.println("\n1. Push a character");
            System.out.println("2. Pop the last input");
            System.out.println("3. Pop the last input in order");
            System.out.println("4. Peek the last input");
            System.out.println("5. Copy last input to clipboard");
            System.out.println("6. Paste from clipboard");
            System.out.println("7. Undo last input");
            System.out.println("8. Quit");
            System.out.print("Enter your choice: ");
            String choice = scanner.nextLine();

            switch (choice) {
                case "1":
                    System.out.print("Enter a character to push: ");
                    String character = scanner.nextLine();
                    stack.push(character);
                    break;
                case "2":
                    popLastInput(stack);
                    break;
                case "3":
                    popLastInputOrdered(stack);
                    break;
                case "4":
                    peekLastInput(stack);
                    break;
                case "5":
                    copyToClipboard(stack, clipboard);
                    break;
                case "6":
                    pasteFromClipboard(stack, clipboard);
                    break;
                case "7":
                    undo(stack);
                    break;
                case "8":
                    System.out.println("Exiting program...");
                    scanner.close();
                    return;
                default:
                    System.out.println("Invalid choice. Please enter a valid option.");
            }
        }
    }

    public static void popLastInput(StackWithTimestamp stack) {
        if (!stack.isEmpty()) {
            StackWithTimestamp.ItemWithTimestamp item = stack.pop();
            System.out.println("Pop the last input: " + item.getItem());
        } else {
            System.out.println("Stack is empty.");
        }
    }

    public static void popLastInputOrdered(StackWithTimestamp stack) {
        if (!stack.isEmpty()) {
            StringBuilder orderedInputs = new StringBuilder();
            while (!stack.isEmpty()) {
                StackWithTimestamp.ItemWithTimestamp item = stack.pop(); // 타임스탬프를 무시하고 아이템만 가져옴
                orderedInputs.append(item.getItem());
            }
            System.out.println("Pop the last input in order: " + orderedInputs.toString());
        } else {
            System.out.println("Stack is empty.");
        }
    }

    public static void peekLastInput(StackWithTimestamp stack) {
        if (!stack.isEmpty()) {
            StackWithTimestamp.ItemWithTimestamp item = stack.peek();
            System.out.println("Peek the last input: " + item.getItem());
        } else {
            System.out.println("Stack is empty.");
        }
    }

    public static void copyToClipboard(StackWithTimestamp stack, StackWithTimestamp clipboard) {
        if (!stack.isEmpty()) {
            StackWithTimestamp.ItemWithTimestamp item = stack.peek();
            clipboard.push(item.getItem());
            System.out.println("Last input '" + item.getItem() + "' copied to clipboard.");
        } else {
            System.out.println("Stack is empty.");
        }
    }

    public static void pasteFromClipboard(StackWithTimestamp stack, StackWithTimestamp clipboard) {
        if (!clipboard.isEmpty()) {
            StackWithTimestamp.ItemWithTimestamp item = clipboard.peek();
            stack.push(item.getItem());
            System.out.println("Last input '" + item.getItem() + "' pasted from clipboard.");
        } else {
            System.out.println("Clipboard is empty.");
        }
    }

    public static void undo(StackWithTimestamp stack) {
        if (!stack.isEmpty()) {
            StackWithTimestamp.ItemWithTimestamp item = stack.pop();
            System.out.println("Undo last input: " + item.getItem());
        } else {
            System.out.println("Stack is empty.");
        }
    }
}

File: ChatGPT(예진)/Cpp (수정 완료)/Stack_Manipulator.cpp
#include <iostream>
#include <vector>
#include <chrono>

struct StackItem {
    char item;
    double timestamp;
};

class Stack {
public:
    Stack(int capacity = 10) : items(), capacity(capacity) {
        items.reserve(capacity);
    }

    void push(char item) {
        if (items.size() == capacity) {
            capacity *= 2;
            items.reserve(capacity);
        }
        double timestamp = static_cast<double>(std::chrono::duration_cast<std::chrono::milliseconds>(
            std::chrono::system_clock::now().time_since_epoch()).count()) / 1000.0;
        items.push_back({ item, timestamp });
    }

    StackItem pop() {
        if (!isEmpty()) {
            StackItem item = items.back();
            items.pop_back();
            return item;
        }
        return { '\0', 0.0 };
    }

    bool isEmpty() const {
        return items.empty();
    }

    StackItem peek() const {
        if (!isEmpty()) {
            return items.back();
        }
        return { '\0', 0.0 };
    }

private:
    std::vector<StackItem> items;
    int capacity;
};

void mainMenu() {
    std::cout << "\n1. Push a character\n";
    std::cout << "2. Pop the last input\n";
    std::cout << "3. Pop the last input in order\n";
    std::cout << "4. Peek the last input\n";
    std::cout << "5. Copy last input to clipboard\n";
    std::cout << "6. Paste from clipboard\n";
    std::cout << "7. Undo last input\n";
    std::cout << "8. Quit\n";
}

void popLastInput(Stack& stack) {
    if (!stack.isEmpty()) {
        StackItem item = stack.pop();
        std::cout << "Pop the last input: " << item.item << '\n';
    }
    else {
        std::cout << "Stack is empty.\n";
    }
}

void popLastInputOrdered(Stack& stack) {
    if (!stack.isEmpty()) {
        std::vector<StackItem> orderedItems;
        while (!stack.isEmpty()) {
            orderedItems.push_back(stack.pop());
        }
        for (auto it = orderedItems.rbegin(); it != orderedItems.rend(); ++it) {
            std::cout << it->item;
        }
        std::cout << '\n';
    }
    else {
        std::cout << "Stack is empty.\n";
    }
}

void peekLastInput(const Stack& stack) {
    if (!stack.isEmpty()) {
        StackItem item = stack.peek();
        std::cout << "Peek the last input: " << item.item << '\n';
    }
    else {
        std::cout << "Stack is empty.\n";
    }
}

void copyToClipboard(const Stack& stack, Stack& clipboard) {
    if (!stack.isEmpty()) {
        StackItem item = stack.peek();
        clipboard.push(item.item);
        std::cout << "Last input '" << item.item << "' copied to clipboard.\n";
    }
    else {
        std::cout << "Stack is empty.\n";
    }
}

void pasteFromClipboard(Stack& stack, const Stack& clipboard) {
    if (!clipboard.isEmpty()) {
        StackItem item = clipboard.peek();
        stack.push(item.item);
        std::cout << "Last input '" << item.item << "' pasted from clipboard.\n";
    }
    else {
        std::cout << "Clipboard is empty.\n";
    }
}

void undo(Stack& stack) {
    if (!stack.isEmpty()) {
        StackItem item = stack.pop();
        std::cout << "Undo last input: " << item.item << '\n';
    }
    else {
        std::cout << "Stack is empty.\n";
    }
}

int main() {
    Stack stack;
    Stack clipboard;

    while (true) {
        mainMenu();
        int choice;
        std::cout << "Enter your choice: ";
        std::cin >> choice;

        if (choice == 1) {
            char character;
            std::cout << "Enter a character to push: ";
            std::cin >> character;
            stack.push(character);
        }
        else if (choice == 2) {
            popLastInput(stack);
        }
        else if (choice == 3) {
            popLastInputOrdered(stack);
        }
        else if (choice == 4) {
            peekLastInput(stack);
        }
        else if (choice == 5) {
            copyToClipboard(stack, clipboard);
        }
        else if (choice == 6) {
            pasteFromClipboard(stack, clipboard);
        }
        else if (choice == 7) {
            undo(stack);
        }
        else if (choice == 8) {
            std::cout << "Exiting program...\n";
            break;
        }
        else {
            std::cout << "Invalid choice. Please enter a valid option.\n";
        }
    }

    return 0;
}
